packages/edit-site/src/components/post-router/index.js

```diff
@@ -0,0 +1,63 @@
+/**
+ * WordPress dependencies
+ */
+import { useEffect } from '@wordpress/element';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { getQueryArg } from '@wordpress/url';
+
+/**
+ * Internal dependencies
+ */
+import updateQueryParams from './update-query-params';
+
+export default function PostRouter() {
+	const { templateId, templatePartId, templateType, page } = useSelect(
+		( select ) => {
+			const {
+				getTemplateId,
+				getTemplatePartId,
+				getTemplateType,
+				getPage,
+			} = select( 'core/edit-site' );
+
+			return {
+				templateId: getTemplateId(),
+				templatePartId: getTemplatePartId(),
+				templateType: getTemplateType(),
+				page: getPage(),
+			};
+		}
+	);
+	const { setTemplate, setTemplatePart, showHomepage, setPage } = useDispatch(
+		'core/edit-site'
+	);
+
+	// Set correct entity on load.
+	useEffect( () => {
+		const url = window.location.href;
+		const contextType = getQueryArg( url, 'contextType' );
+		const id = getQueryArg( url, 'id' );
+		const content = getQueryArg( url, 'content' );
+
+		if ( 'content' === contextType && content ) {
+			setPage( JSON.parse( content ) );
+		} else if ( 'wp_template' === contextType && id ) {
+			setTemplate( id );
+		} else if ( 'wp_template_part' === contextType && id ) {
+			setTemplatePart( id );
+		} else {
+			showHomepage();
+		}
+	}, [] );
+
+	// Update URL when context changes.
+	useEffect( () => {
+		if ( page ) {
+			updateQueryParams( 'content', JSON.stringify( page ) );
```

Comment:
I'm not sure if we should pass all of the page object details into the URL. It seems to me that most of the data there is redundant and can be inferred by post id. For example:

```js
{"type":"page","slug":"sample-page","path":"/?page_id=2","context":{"postType":"page","postId":2}}
```

All of the features outside of editor that want to integrate this will have to conform with this format. They will become and every change to this might break all of the integrations. I think we should try to only pass necessary data as separate URL params, instead of URL encoding an exiting object.

Reply:
Good Points.  My main reason for using the entire page object was so we could easily support every editor state, including taxonomies/categories/etc.  We can still develop support for these without setting the entire page object as the value, but we I think we will need to add more individual query args in that case and rebuild the page object from them.